runners/direct-java/src/main/java/org/apache/beam/runners/direct/DoFnLifecycleManager.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.direct;
+
+import org.apache.beam.sdk.runners.PipelineRunner;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.OldDoFn;
+import org.apache.beam.sdk.util.SerializableUtils;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+
+/**
+ * Manages {@link DoFn} setup, teardown, and serialization.
+ *
+ * <p>{@link DoFnLifecycleManager} is similar to a {@link ThreadLocal} storing a {@link DoFn}, but
+ * calls the {@link DoFn} {@link Setup} the first time the {@link DoFn} is obtained and {@link
+ * Teardown} whenever the {@link DoFn} is removed, and provides a method for clearing all cached
+ * {@link DoFn DoFns}.
+ */
+class DoFnLifecycleManager {
+  private static final Logger LOG = LoggerFactory.getLogger(DoFnLifecycleManager.class);
+
+  public static DoFnLifecycleManager of(OldDoFn<?, ?> original) {
+    return new DoFnLifecycleManager(original);
+  }
+
+  private final LoadingCache<Thread, OldDoFn<?, ?>> outstanding;
+
+  private DoFnLifecycleManager(OldDoFn<?, ?> original) {
+    this.outstanding = CacheBuilder.newBuilder().build(new DeserializingCacheLoader(original));
+  }
+
+  public OldDoFn<?, ?> get() throws Exception {
```

Comment:
Should this be modified to just store/execute new `DoFn` rather than `OldDoFn`? Assuming `OldDoFn` will be removed soon, this seems like it should support the new thing.